src/System.Linq.Expressions/src/System/Linq/Expressions/BinaryExpression.cs

```diff
@@ -266,25 +266,25 @@ private Expression ReduceIndex()
 
             var index = (IndexExpression)_left;
 
-            var vars = new List<ParameterExpression>(index.ArgumentCount + 2);
-            var exprs = new List<Expression>(index.ArgumentCount + 3);
```

Comment:
This is exactly the case that turns out to be more expensive than it should be. We know the size upfront, yet use a `List<T>` so we don't have to track an index into the array to assign to next (which is non-trivial due to the branching below). We then went on to pass the `List<T>` to the `Block` factory as-is, causing a copy to be made in its call to `ToReadOnly`. Given these collections exhibit ownership transfer behavior, we should be able to just create a `TrueReadOnlyCollection<T>` wrapper around a `T[]` and avoid any copies.